README.md
# lua-resty-txid

[![CircleCI](https://circleci.com/gh/GUI/lua-resty-txid.svg?style=svg)](https://circleci.com/gh/GUI/lua-resty-txid)

lua-resty-txid provides a function that can be used to generate unique transaction/request IDs for [OpenResty/nginx](http://openresty.org). The IDs can be used to correlate logs or upstream requests and have the following characteristics:

- 20 characters
- base32hex encoded
- Temporally and lexically sortable
- Case insensitive
- 96 bit identifier

lua-resty-txid is a LuaJIT port of [ngx\_txid](https://github.com/streadway/ngx_txid) for [OpenResty](https://openresty.org/) (or nginx with [ngx\_lua](https://github.com/openresty/lua-nginx-module#installation)). The IDs generated by lua-resty-txid follow the exact same pattern and are compatible with ngx\_txid.

## Installation

Via [OPM](https://opm.openresty.org):

```sh
opm get GUI/lua-resty-txid
```

Or via [LuaRocks](https://luarocks.org):

```sh
luarocks install lua-resty-txid
```

## Usage

A single `txid()` Lua function is exposed by this module to generate IDs:

```lua
local txid = require "resty.txid"
local id = txid() -- b2g6q94qdn6h84an7vfg
```

Each time `txid()` is called, a new, unique ID will be returned, so you will need to cache the result if you wish to reuse the same ID in multiple places for a single request. Depending on your usage, [`ngx.ctx`](https://github.com/openresty/lua-nginx-module#ngxctx) or [`set_by_lua`](https://github.com/openresty/lua-nginx-module#set_by_lua) offer some simple options for caching the value on a per-request basis.

```lua
txid() -- b2g83t2oshrg092mjggg
txid() -- b2g83t2oodncokuges00

ngx.ctx.txid = txid() -- b2g83t2od939mdvb2l0g
ngx.ctx.txid          -- b2g83t2od939mdvb2l0g
```

Finally, `txid()` accepts an optional argument for what timestamp (in milliseconds) to use when generating the ID. By default, the current timestamp is used. Since the resulting IDs are temporally and lexically sortable, this can be used to generate IDs that will be sorted based on a previous date or time.

```lua
local timestamp_ms = 655829050000 -- 1990-10-13 14:44:10
txid(timestamp_ms) -- 4om9qi54la8ffr4bd9sg

local timestamp_ms = 655929050000 -- 1990-10-14 12:30:50
txid(timestamp_ms) -- 4on1lg74nt0ud2ssllu0
```

### Example

A more complete example, with caching, setting request/response headers, and integration with nginx's logging:

```nginx
http {
  log_format agent "$lua_txid $http_user_agent";
  log_format addr "$lua_txid $remote_addr";

  init_by_lua_block {
    # Pre-load the module.
    require "resty.txid"
  }

  server {
    listen 8080;
    access_log logs/agents.log agent;
    access_log logs/addrs.log addr;

    # Set an nginx variable that is cached per request and can be used in the
    # nginx log_format.
    set_by_lua_block $lua_txid {
      local txid = require "resty.txid"
      return txid()
    }

    location / {
      # Set a header on the response providing the ID.
      more_set_headers "X-Request-Id: $lua_txid";

      # Set a header on the request providing the ID (which will be sent to the
      # proxied upstream).
      more_set_input_headers "X-Request-Id: $lua_txid";

      proxy_pass http://localhost:8081;
    }
  }
}
```

## Performance

Benchmarks indicate that performance is equivalent to the [ngx\_txid](https://github.com/streadway/ngx_txid) C extension.

## Design

The transaction ID design is a direct port of [ngx\_txid](https://github.com/streadway/ngx_txid), so here's all the original information about the design from ngx\_txid:

### Background

The design of this transaction ID should meet the following requirements:

- Be roughly numerically temporally sortable with ~second granularity.
- Have a representation that is roughly lexically sortable with ~second granularity.
- Have a probability of less than 1e-9 for collision at 1 million transactions per second.
- Be efficient and easy to decode into fixed size C types
- Always be available at the risk of higher collision probability
- Use as few bytes as possible
- Work with IPv4 and IPv6 networks

### Technique

Use a monotonic millisecond resolution clock in the high 42 bits and system entropy for the low 54 bits. Use enough entropy bits to satisfy a collision probability at a desired global request rate.

```
+------------- 64 bits------------+--- 32 bits ----+
+------ 42 bits ------+--22 bits--|----------------+
| msec since 1970-1-1 | random    | random         |
+---------------------+-----------+----------------+
```

A request rate of 1 million per second across all servers means 1000 random values per millisecond.  Estimating the collision probability using the [birthday paradox](http://en.wikipedia.org/wiki/Birthday_problem) can be done with this formula: `1 - e^(-((m^2)/(2*n)))` where `m` is the number of ids and `n` is the number of random values possible.

When using 54 bits of entropy:

```
1mil req/s  = 1 - exp(-((1000^2) /(2*2^54))) = 2.775558e-11
10mil req/s = 1 - exp(-((10000^2)/(2*2^54))) = 2.775558e-09
```

The odds of collision are small even at 10 million requests per second.

Nginx keeps track of the current clock in increments of the configuration directive `timer_resolution`.  The clock resolution for `$txid` is 1ms, so a timer resolution greater than 1ms means that the probability of collision will increase.  If you have a `timer_resolution` of 10ms, 1 million requests per second would require 10,000 random values per second in the worst case.

### Encoding

[base32hex](https://en.wikipedia.org/wiki/Base32#base32hex) is used with a lower case alphabet and without padding characters is chosen for the following reasons:

- Lexically sort order equivalent to numeric sort order
- Case insensitive equality
- Lower case is easer for visual compares
- Denser than hex encoding by 4 bytes

### Other techniques

- [snowflake](https://github.com/twitter/snowflake): Uses time(41) + unique id(10) + sequence(12).
  - Pro: Guaranteed unique sequences
  - Pro: Fits in 63 bits
  - Cons: Requires unique id coordination for each server - 16 workers processes per host means a limit of 64 instances of nginx
  - Cons: Only 11 bits available for unique id, needs monitoring
  - Cons: Total ordering only possible in the same process
  - Cons: Service interruption possible when clocks lose synchronization

- [flake](https://github.com/boundary/flake): Uses time + mac id + sequence.
  - Pro: Guaranteed unique sequences
  - Cons: Uses 128 bits
  - Cons: Wastes 22 bits of timestamp data
  - Cons: Only a single process per host can generate ids - needs to synchronize access to the sequence from each worker process
  - Cons: Service interruption possible when clocks lose synchronization
  - Cons: Seeds cross platform MAC Address lookup.

- [UUIDv4](http://www.ietf.org/rfc/rfc4122.txt): 122 bits of entropy
  - Pro: Very low probability of collision
  - Cons: Unsortable

- [UUID with timestamp](http://www.ietf.org/rfc/rfc4122.txt): 48 bits of time + 74 bits entropy
  - Pro: Very low probability of collision
  - Cons: String representation is not temporally local

- [httpd mod\_unique\_id](http://httpd.apache.org/docs/2.4/mod/mod_unique_id.html): Host ip(32) + pid(32) + time(32) + sequence (16) + thread id (32)
  - Pro: Deterministic
  - Cons: Uses 144 bits
  - Cons: Assumes unique IPv4 for the hostnamme's interface
  - Cons: Unsortable case-sensitive custom representation - base64 with a custom alphabet
  - Cons: Hard limit of 65535 ids per second per pid - small tolerance for clock steps

## Development

After checking out the repo, Docker can be used to run the test suite:

```sh
docker-compose run --rm app make test
```

### Release Process

To publish releases to OPM and LuaRocks:

```sh
VERSION=x.x.x make release
```

## Credits

Credit for the original design goes to [ngx\_txid](https://github.com/streadway/ngx_txid). This is simply a pure LuaJIT port for easier installation in OpenResty.


CHANGELOG.md
## lua-resty-txid Change Log

## 1.0.0 - 2018-04-01

- Initial release.


spec/txid_spec.lua
-- Note that we loop over many tests 100 times when checking timestamp behavior
-- to ensure that the first 8 chars only correspond to the timestamp and don't
-- contain any of the random bits.
describe("txid", function()
  local txid = require "resty.txid"

  it("generates a 20 character base32hex encoded id", function()
    local id = txid()
    assert.truthy(ngx.re.match(id, [[^[0-9a-v]{20}$]]), id)
  end)

  it("generates ids that are unique", function()
    local ids = {}
    for _ = 1, 100 do
      local id = txid()
      assert.truthy(ngx.re.match(id, [[^[0-9a-v]{20}$]]), id)

      assert.is_nil(ids[id])

      ids[id] = 1
      assert.is_not_nil(ids[id])
    end
  end)

  it("generates ids at the same timestamp that are unique", function()
    local ids = {}
    for _ = 1, 100 do
      local id = txid(1522545688552) -- 2018-04-01 01:21:28
      assert.truthy(ngx.re.match(id, [[^b2fr5nvq[0-9a-v]{12}$]]), id)

      assert.is_nil(ids[id])

      ids[id] = 1
      assert.is_not_nil(ids[id])
    end
  end)

  it("generates ids that are lexically sortable", function()
    local ids = {}
    for _ = 1, 10 do
      -- Briefly sleep between ID generation to ensure the initial timestamp
      -- portion of the id will increase.
      ngx.sleep(0.1)

      table.insert(ids, txid())
    end

    local original_order = { unpack(ids) }
    local sorted_order = { unpack(ids) }
    table.sort(sorted_order)

    assert.equal(10, #sorted_order)
    assert.same(sorted_order, original_order)
  end)

  it("generates expected ids in the past", function()
    for _ = 1, 100 do
      local id = txid(655829050000) -- 1990-10-13 14:44:10
      assert.truthy(ngx.re.match(id, [[^4om9qi54[0-9a-v]{12}$]]), id)
    end
  end)

  it("generates expected ids in the future", function()
    for _ = 1, 100 do
      local id = txid(4387530288000) -- 2109-01-13 14:24:48
      assert.truthy(ngx.re.match(id, [[^vthknin0[0-9a-v]{12}$]]), id)
    end
  end)

  it("generates ids that are lexically sortable within a few milliseconds precision", function()
    for _ = 1, 100 do
      local id = txid(0) -- 1970-01-01 00:00:00.000
      assert.truthy(ngx.re.match(id, [[^00000000[0-9a-v]{12}$]]), id)

      -- The last 2 bits of the timestamp are mixed with random data, so it
      -- requires at least a 4 milliseconds increment the portion of the ID
      -- that is reliably lexically sortable.
      local id_next = txid(4) -- 1970-01-01 00:00:00.004
      assert.truthy(ngx.re.match(id_next, [[^00000001[0-9a-v]{12}$]]), id_next)
    end
  end)

  it("generates expected ids at the epoch boundary", function()
    for _ = 1, 100 do
      local id_at_epoch = txid(0) -- 1970-01-01 00:00:00.000
      assert.truthy(ngx.re.match(id_at_epoch, [[^00000000[0-9a-v]{12}$]]), id_at_epoch)

      local id_before_epoch = txid(-1) -- 1969-12-31 23:59:59.999
      assert.truthy(ngx.re.match(id_before_epoch, [[^vvvvvvvv[0-9a-v]{12}$]]), id_before_epoch)
    end
  end)


  it("generates expected ids at the 42 bit timestamp boundary", function()
    for _ = 1, 100 do
      local id_at_max = txid(4398046511103) -- 2109-05-15 07:35:11.103
      assert.truthy(ngx.re.match(id_at_max, [[^vvvvvvvv[0-9a-v]{12}$]]), id_at_max)

      local id_after_max = txid(4398046511104) -- 2109-05-15 07:35:11.104
      assert.truthy(ngx.re.match(id_after_max, [[^00000000[0-9a-v]{12}$]]), id_after_max)
    end
  end)

  it("rolls over the sortable timestamp representation at the epoch and 42 bit timestamp maximum", function()
    for _ = 1, 100 do
      -- The same time prior to the epoch and prior to the 42bit timestamp
      -- maximum should have equivalent representations.
      local id_before_epoch = txid(-1000000000) -- 1969-12-20 10:13:20
      assert.truthy(ngx.re.match(id_before_epoch, [[^vvohijc0[0-9a-v]{12}$]]), id_before_epoch)

      local id_before_max = txid(4398046511104 - 1000000000) -- 2109-05-03 17:48:31
      assert.truthy(ngx.re.match(id_before_max, [[^vvohijc0[0-9a-v]{12}$]]), id_before_max)

      -- Similarly, the same time after the epoch and after the 42bit timestamp
      -- maximum should appear similarly (the 42bit timestamp maximum becomes
      -- the new epoch).
      local id_after_epoch = txid(1000000000) -- 1970-01-12 13:46:40
      assert.truthy(ngx.re.match(id_after_epoch, [[^007edck0[0-9a-v]{12}$]]), id_after_epoch)

      local id_after_max = txid(4398046511104 + 1000000000) -- 2109-05-26 21:21:51
      assert.truthy(ngx.re.match(id_after_max, [[^007edck0[0-9a-v]{12}$]]), id_after_max)
    end
  end)
end)


lib/resty/txid/base32.lua
local bitwise = require "bit"

local bitwise_and = bitwise.band
local bitwise_lshift = bitwise.lshift
local bitwise_or = bitwise.bor
local bitwise_rshift = bitwise.rshift

local BYTE_MASK = 0x1f
local BASE32_ALPHABET = "0123456789abcdefghijklmnopqrstuv"
local BASE32_ALPHABET_TABLE = {}
for index = 1, #BASE32_ALPHABET do
  local char = BASE32_ALPHABET:sub(index, index)
  BASE32_ALPHABET_TABLE[index - 1] = char
end

return function(src, pad)
  local out = {}
  local len = #src
  local offset = 0
  while len > 0 do
    local bit1 = 0
    local bit2 = 0
    local bit3 = 0
    local bit4 = 0
    local bit5 = 0
    local bit6 = 0
    local bit7 = 0
    local bit8 = 0

    if len >= 5 then
      local byte5_pos = 5 + offset
      local src_byte5 = string.byte(string.sub(src, byte5_pos, byte5_pos))
      bit8 = bitwise_or(bit8, bitwise_and(src_byte5, BYTE_MASK))
      bit7 = bitwise_or(bit7, bitwise_rshift(src_byte5, 5))
    end

    if len >= 4 then
      local byte4_pos = 4 + offset
      local src_byte4 = string.byte(string.sub(src, byte4_pos, byte4_pos))
      bit7 = bitwise_or(bit7, bitwise_and(bitwise_lshift(src_byte4, 3), BYTE_MASK))
      bit6 = bitwise_or(bit6, bitwise_and(bitwise_rshift(src_byte4, 2), BYTE_MASK))
      bit5 = bitwise_or(bit5, bitwise_rshift(src_byte4, 7))
    end

    if len >= 3 then
      local byte3_pos = 3 + offset
      local src_byte3 = string.byte(string.sub(src, byte3_pos, byte3_pos))
      bit5 = bitwise_or(bit5, bitwise_and(bitwise_lshift(src_byte3, 1), BYTE_MASK))
      bit4 = bitwise_or(bit4, bitwise_and(bitwise_rshift(src_byte3, 4), BYTE_MASK))
    end

    if len >= 2 then
      local byte2_pos = 2 + offset
      local src_byte2 = string.byte(string.sub(src, byte2_pos, byte2_pos))
      bit4 = bitwise_or(bit4, bitwise_and(bitwise_lshift(src_byte2, 4), BYTE_MASK))
      bit3 = bitwise_or(bit3, bitwise_and(bitwise_rshift(src_byte2, 1), BYTE_MASK))
      bit2 = bitwise_or(bit2, bitwise_and(bitwise_rshift(src_byte2, 6), BYTE_MASK))
    end

    if len >= 1 then
      local byte1_pos = 1 + offset
      local src_byte1 = string.byte(string.sub(src, byte1_pos, byte1_pos))
      bit2 = bitwise_or(bit2, bitwise_and(bitwise_lshift(src_byte1, 2), BYTE_MASK))
      bit1 = bitwise_or(bit1, bitwise_rshift(src_byte1, 3))
    end

    if len >= 1 then
      table.insert(out, BASE32_ALPHABET_TABLE[bit1])
      table.insert(out, BASE32_ALPHABET_TABLE[bit2])
    elseif pad then
      table.insert(out, "=")
      table.insert(out, "=")
    end

    if len >= 2 then
      table.insert(out, BASE32_ALPHABET_TABLE[bit3])
      table.insert(out, BASE32_ALPHABET_TABLE[bit4])
    elseif pad then
      table.insert(out, "=")
      table.insert(out, "=")
    end

    if len >= 3 then
      table.insert(out, BASE32_ALPHABET_TABLE[bit5])
      table.insert(out, BASE32_ALPHABET_TABLE[bit6])
    elseif pad then
      table.insert(out, "=")
      table.insert(out, "=")
    end

    if len >= 4 then
      table.insert(out, BASE32_ALPHABET_TABLE[bit7])
      table.insert(out, BASE32_ALPHABET_TABLE[bit8])
    elseif pad then
      table.insert(out, "=")
      table.insert(out, "=")
    end

    len = len - 5
    offset = offset + 5
  end

  local out_str = table.concat(out, "")
  return out_str
end



lib/resty/txid.lua
local base32hex = require "resty.txid.base32hex"
local bitwise = require "bit"
local ffi = require "ffi"
local resty_random = require "resty.random"

local bitwise_and = bitwise.band
local bitwise_lshift = bitwise.lshift
local bitwise_or = bitwise.bor
local bitwise_rshift = bitwise.rshift
local ffi_cast = ffi.cast

local BYTE_SIZE = 8
local TIMESTAMP_SIZE = 64
local TRUNCATED_TIMESTAMP_SIZE = 42
local TRUNCATED_TIMESTAMP_SHIFT = TIMESTAMP_SIZE - TRUNCATED_TIMESTAMP_SIZE
local BYTE_MASK = 0xff

local BYTE1_SHIFT = TIMESTAMP_SIZE - 1 * BYTE_SIZE
local BYTE2_SHIFT = TIMESTAMP_SIZE - 2 * BYTE_SIZE
local BYTE3_SHIFT = TIMESTAMP_SIZE - 3 * BYTE_SIZE
local BYTE4_SHIFT = TIMESTAMP_SIZE - 4 * BYTE_SIZE
local BYTE5_SHIFT = TIMESTAMP_SIZE - 5 * BYTE_SIZE
local BYTE6_SHIFT = TIMESTAMP_SIZE - 6 * BYTE_SIZE

local function byte_from_time(msec, shift)
  return tonumber(bitwise_and(bitwise_rshift(msec, shift), BYTE_MASK))
end

return function(time)
  if not time then
    time = ngx.now() * 1000
  end

  -- Generate random data to append after the timestamp.
  local random = resty_random.bytes(7)

  -- Convert the time to a 64 bit integer for further manipulation.
  local msec = ffi_cast("uint64_t", time)

  -- The timestamp is 64 bits, but shorten it to 42 bits. This is enough to
  -- store dates up to 2109-05-15 (4398046511103 milliseconds past epoch)
  -- before cycling back to the original epoch values.
  local msec_truncated = bitwise_lshift(msec, TRUNCATED_TIMESTAMP_SHIFT)

  -- Extract individual bytes from the timestamp as numeric character codes.
  local byte1 = byte_from_time(msec_truncated, BYTE1_SHIFT)
  local byte2 = byte_from_time(msec_truncated, BYTE2_SHIFT)
  local byte3 = byte_from_time(msec_truncated, BYTE3_SHIFT)
  local byte4 = byte_from_time(msec_truncated, BYTE4_SHIFT)
  local byte5 = byte_from_time(msec_truncated, BYTE5_SHIFT)

  -- For the 6th byte, use the last 2 bits from the timestamp (bits 41-42), and
  -- the rest of the byte is shared with the first byte of random data.
  local byte6 = bitwise_or(byte_from_time(msec_truncated, BYTE6_SHIFT), string.byte(string.sub(random, 1, 1)))

  -- Construct the full ID by converting the timestamp character codes to their
  -- string values (6 bytes, 48 total bits, 42 of which represent the
  -- timestamp), and then append the rest of the random data (6 bytes/48 bits),
  -- resulting in a 12 byte/96 bit ID.
  local data = string.char(byte1, byte2, byte3, byte4, byte5, byte6) .. string.sub(random, 2)

  -- Base32hex encode the binary data (without padding), resulting in a 20
  -- character, lexically sortable sting.
  local encoded = base32hex(data, false)
  return encoded
end
